### DISCUSSION TITLE
Sysex dump to modulator value

### DISCUSSION THREAD

**Discussion by jojogib** [RomanKubiak/ctrlr#635] [Q&A]
Hey, 
I just finished a panel but i can´t figure out how to translate the received sysex dump msg to a modulator value.
When i press "Get Sysex Dump"  I receive this msg:

f0 42 21 06 00 00 01  00 01 00 07 00 00 00 00 00 00 01 0f 00 00 00 00 00 01 00 01 00 00 00 00 00 00 01 0f 00 00 00 00 00 02 00 00 00 00 00 00 00 00 00 00 00 00 05 07 00 00 00 00 00 06 00 00 00 00 00 00 00 00 00 00 01 0f 00 00 00 01 01 0b 00 00 00 02 00 00 00 00 00 0f 00 0e 01 0b 00 04 00 00 00 00 00 00 01 05 01 09 00 0b 00 00 00 09 01 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 07 00 02 00 00 00 00 00 00 00 00 00 00 00 00 00 07 00 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 03 0f 00 00 00 00 00 00 f7

When I change octave I get this msg:

f0 42 21 06 00 00 02  00 01 00 07 00 00 00 00 00 00 01 0f 00 00 00 00 00 01 00 01 00 00 00 00 00 00 01 0f 00 00 00 00 00 02 00 00 00 00 00 00 00 00 00 00 00 00 05 07 00 00 00 00 00 06 00 00 00 00 00 00 00 00 00 00 01 0f 00 00 00 01 01 0b 00 00 00 02 00 00 00 00 00 0f 00 0e 01 0b 00 04 00 00 00 00 00 00 01 05 01 09 00 0b 00 00 00 09 01 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 07 00 02 00 00 00 00 00 00 00 00 00 00 00 00 00 07 00 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 03 0f 00 00 00 00 00 00 f7

So octave is realted to:

f0 42 21 06 00 **00 02**

I think I need LUA script 

When midi in is receive this dump msg get byte 6,7 and ssend it to the modulator?
Thanks for help [syxexDumpToModulator.bpanelz.zip](https://github.com/RomanKubiak/ctrlr/files/15068116/syxexDumpToModulator.bpanelz.zip)

**Comment by dnaldoog** [comment]
You need a function that runs when the panel receives a midi message:

Place it here:
![image](https://github.com/RomanKubiak/ctrlr/assets/76831683/a1b4597f-5caa-4102-94b7-9fa389b22eeb)

In the midi received function something like:
<pre>
<code>
--
-- Called when a panel receives a midi message (does not need to match any modulator mask)
-- @midi   CtrlrMidiMessage object
--

myMidiReceived = function(--[[ CtrlrMidiMessage --]] midi)
    local s = midi:getSize()
    if s == 262 then
        local octaveMsb = midi:getData():getByte(5)
        local octaveLsb = midi:getData():getByte(6)
        local sum = x(octaveMsb, octaveLsb)
        -- two bytes could represent 4 bit nibbles or 14 bit data

        panel:getComponent("Octave"):setValue(sum, true)
    end
end

x = function(m, l)
    -- two bytes could represent 4 bit nibbles or 14 bit data???
    -- Here I'll use 4 bit data
   -- If 14 bit, something like m=m*128
    m = bit.lshift(m, 4)
    return m + l
end

</code>
</pre>
----
[Ctrlr Panel4_1.zip](https://github.com/RomanKubiak/ctrlr/files/15069341/Ctrlr.Panel4_1.zip)

**Reply by jojogib** [reply]
Hey thanks for the reply and your help, It works .. kind of. If the moldulator is above 15 I get weird results.
15 = 00 0f
16= 01 00  

 -- If 14 bit, something like m=m*128.   how do I implemnt this exactly?
thank you

**Comment by dnaldoog** [comment]
Yes, so in 14 bit if a 2 byte number comes in, you just multiply the MSByte by 128 (as long as this incoming is not representing negative numbers; if it is let me know) - you add that to the LSB and you have the value to assign to a modulator.

Example 01 02 = (128 * 1) + 2 => 130

Going the other way to create your sysex, you need to split the number into two bytes, 

msbforsysex=bit,rshift(130,7) -- = 01 (byte 3 below)
lsbforsysex=bit,band(130,0x7f) -- = 02 (byte 4 below)

`f0 00 00 01 02 f7`

**Comment by dnaldoog** [comment]
I didn't read that thoroughly. What you have there are 4 bit nibbles. For example 0f 0f would be ff which is 255.

The best way is to run the two bytes through a function

<pre>
function combine4n(a, b) -- convert two 4 bit nibbles to a single integer
    local m = bit.lshift(a, 4)
    return m + b
end
-- or

function combine4n(a, b) -- convert two 4 bit nibbles to a single integer
    return bit.bor(bit.lshift(a, 4), b)
end

</pre>

if you pass 1,0 to this function it will return 16

**Comment by jojogib** [comment]
Thank you I think I got it.
